src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -462,11 +443,47 @@ HRESULT STDMETHODCALLTYPE CorProfiler::JITCompilationStarted(
 
     if (!module_metadata->TryGetWrapperMemberRef(wrapper_method_key,
                                                  wrapper_method_ref)) {
-      // no method ref token found for wrapper method, we can't do the
-      // replacement, this should never happen because we always try to
-      // add the method ref in ModuleLoadFinished()
-      // TODO: log this
-      return S_OK;
+      const auto module_info = GetModuleInfo(this->info_, module_id);
+      if (!module_info.IsValid()) {
+        continue;
+      }
+
+      mdModule module;
+      hr = module_metadata->metadata_import->GetModuleFromScope(&module);
+      if (FAILED(hr)) {
+        Warn("JITCompilationStarted failed to get module metadata token for module_id=",
+             module_id, " module_name=", module_info.assembly.name, " function_id=", function_id,
+             " token=", function_token, " name=", caller.type.name, ".",
+             caller.name, "()");
+        continue;
+      }
+
+      const MetadataBuilder metadata_builder(*module_metadata, module,
+                                             module_metadata->metadata_import, module_metadata->metadata_emit,
+                                             module_metadata->assembly_import, module_metadata->assembly_emit);
+
+      // for each wrapper assembly, emit an assembly reference
+      hr = metadata_builder.EmitAssemblyRef(
```

Comment:
Do we even need to emit before knowing we are going to replace anything?

Reply:
Would it be reasonable to store a flag on the ModuleMetadata stating whether we have emitted the necessary refs?

Reply:
I'm thinking about the potential case where we may fail to emit or store refs for some reason, should we keep trying for every single JIT?

Reply:
That's a good point, we should be able to defer that until we are about to replace it! I'll add that to this PR.

Reply:
I didn't refresh so I only saw your first comment when responding.

I like the idea of only attempting to emit the refs once, so I'll look into implementing that. Originally we stored the ref only once during `ModuleLoadFinished` and then just retrieved it in `JITCompilationStarted` so that wasn't an issue. We would just exit nicely if we couldn't find it.

Reply:
Implemented in 3aa43a974f95d25712fe8af5d728008a5409afbc

Reply:
Looks great!